sync_check/util/diff.go

```diff
@@ -0,0 +1,577 @@
+		eq, err = df.EqualIndex(tblName)
+		if err != nil {
+			err = errors.Trace(err)
+			return
+		}
+		if !eq {
+			log.Infof("table have different index: %s\n", tblName)
+			equal = false
+		}
+
+		eq, err = df.EqualTable(tblName)
+		if err != nil || !eq {
+			err = errors.Trace(err)
+			equal = false
+			return
+		}
+	}
+
+	return
+}
+
+// EqualTable tests whether two database table have same data and schema.
+func (df *Diff) EqualTable(tblName string) (bool, error) {
+	eq, err := df.equalCreateTable(tblName)
+	if err != nil {
+		return eq, errors.Trace(err)
+	}
+	if !eq {
+		log.Infof("table have different schema: %s\n", tblName)
+		return eq, err
+	}
+
+	eq, err = df.equalTableData(tblName)
+	if err != nil {
+		return eq, errors.Trace(err)
+	}
+	if !eq {
+		log.Infof("table data different: %s\n", tblName)
+	}
+	return eq, err
+}
+
+// EqualIndex tests whether two database index are same.
+func (df *Diff) EqualIndex(tblName string) (bool, error) {
+	index1, err := getTableIndex(df.db1, tblName)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	defer index1.Close()
+	index2, err := getTableIndex(df.db2, tblName)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	defer index2.Close()
+
+	eq, err := equalRows(index1, index2, &showIndex{}, &showIndex{})
+	if err != nil || !eq {
+		return eq, errors.Trace(err)
+	}
+	return eq, nil
+}
+
+func (df *Diff) equalCreateTable(tblName string) (bool, error) {
+	_, err1 := getCreateTable(df.db1, tblName)
+	_, err2 := getCreateTable(df.db2, tblName)
+
+	if errors.IsNotFound(err1) && errors.IsNotFound(err2) {
+		return true, nil
+	}
+	if err1 != nil {
+		return false, errors.Trace(err1)
+	}
+	if err2 != nil {
+		return false, errors.Trace(err2)
+	}
+
+	// TODO ignore table schema currently
+	// return table1 == table2, nil
+	return true, nil
+}
+
+func (df *Diff) equalTableData(tblName string) (bool, error) {
+	dumpJobs, err := generateDumpJob(df.db1, df.dbName, tblName, df.timeField, df.beginTime, df.endTime, df.splitField, df.chunkSize, df.sample)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	checkNums := len(dumpJobs) * df.sample / 100
+	if checkNums == 0 {
+		checkNums = 1
+	}
+	checkNumArr := getRandomN(len(dumpJobs), checkNums)
+	log.Infof("total has %d check jobs, check %+v", len(dumpJobs), checkNumArr)
+
+	checkResultCh := make(chan bool, df.checkThCount)
+	defer close(checkResultCh)
+
+	for i := 0; i < df.checkThCount; i++ {
+		go df.checkChunkDataEqual(dumpJobs, checkNumArr[checkNums*i/df.checkThCount:checkNums*(i+1)/df.checkThCount], tblName, checkResultCh)
+	}
+
+	num := 0
+	for {
+		select {
+		case eq := <-checkResultCh:
+			num++
+			if !eq {
+				return false, nil
+			}
+			if num == df.checkThCount {
+				return true, nil
+			}
+		}
+	}
+}
+
+func (df *Diff) checkChunkDataEqual(dumpJobs []*dumpJob, checkNumArr []int, tblName string, checkResultCh chan bool) {
+	log.Infof("check array: %+v", checkNumArr)
+	if len(checkNumArr) == 0 {
+		checkResultCh <- true
```

Comment:
return true/false as the function's returnning  value
fill it into channel outside base on the determination of calling side